scrapy/resolver.py

```diff
@@ -1,32 +1,49 @@
-from twisted.internet import defer
-from twisted.internet.base import ThreadedResolver
+from twisted.internet import reactor
+from twisted.internet.interfaces import IHostnameResolver, IResolutionReceiver
+from zope.interface.declarations import implementer, provider
 
 from scrapy.utils.datatypes import LocalCache
 
-# TODO: cache misses
 
+# TODO: cache misses
 dnscache = LocalCache(10000)
 
 
-class CachingThreadedResolver(ThreadedResolver):
-    def __init__(self, reactor, cache_size, timeout):
-        super(CachingThreadedResolver, self).__init__(reactor)
-        dnscache.limit = cache_size
+@implementer(IHostnameResolver)
+class CachingHostnameResolver(object):
+
+    def __init__(self, resolver, cache_size, timeout):
+        self.resolver = resolver
         self.timeout = timeout
+        dnscache.limit = cache_size
+
+    def resolveHostName(self, resolutionReceiver, hostName, portNumber=0,
+                        addressTypes=None, transportSemantics='TCP'):
+
+        @provider(IResolutionReceiver)
+        class CachingResolutionReceiver(resolutionReceiver):
+
+            def __init__(self, timeout):
+                self.timeout = timeout
+
+            def resolutionBegan(self, resolution):
+                super(CachingResolutionReceiver, self).resolutionBegan(resolution)
+                self.resolution = resolution
+                # reactor.callLater(self.timeout, resolution.cancel)
```

Comment:
This doesn't really work, [`twisted.internet._resolver.HostResolution`](https://twistedmatrix.com/documents/18.9.0/api/twisted.internet._resolver.HostResolution.html) [doesn't have a `cancel` method](https://github.com/twisted/twisted/blob/twisted-18.9.0/src/twisted/internet/_resolver.py#L31-L41), even though it's supposed to implement [`twisted.internet.interfaces.IHostResolution`](https://twistedmatrix.com/documents/18.9.0/api/twisted.internet.interfaces.IHostResolution.html)

Reply:
https://twistedmatrix.com/trac/ticket/9748